src/repr/src/row.rs

```diff
@@ -2484,8 +2462,9 @@ impl RowPacker<'_> {
         // Validate that what was written maintains the correct invariants.
         let mut actual_datums = 0;
         let mut seen = None;
-        while datum_check < self.row.data.len() {
-            let d = unsafe { read_datum(&self.row.data, &mut datum_check) };
+        let mut dataz = &self.row.data[datum_check..];
+        while !dataz.is_empty() {
+            let d = unsafe { read_datum(&mut dataz) };
```

Comment:
You could convert this to a for-loop over a RowRef on `&self.row.data[datum_check..]`. It needs to yield exactly `expected_datums` elements. Also, remove the `datum_check`, which is now an incorrect assertion (it used to be incremented, but isn't anymore).

Reply:
I deleted the check, which by its own admission would be testing undefined behavior.